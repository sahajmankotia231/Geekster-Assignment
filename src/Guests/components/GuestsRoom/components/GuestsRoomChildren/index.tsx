import React, { ReactElement } from "react";
import { useSelector, useDispatch } from "react-redux";

import Button from "components/Button";
import Select, { Option } from "components/Select";
import { ReactComponent as IconPlus } from "assets/plus.svg";
import { ReactComponent as IconMinus } from "assets/minus.svg";
import {
  StoreStateType,
  RoomType,
  addChild,
  removeChild,
  removeSpecificChild,
  changeChildAge,
} from "store";

import {
  Guest,
  GuestText,
  GuestButtonGroup,
  KidsContainer,
  IconRemoveChild,
} from "../../styles";

type GuestsRoomChildrenProps = {
  position: number;
};

const GuestsRoomChildren = ({
  position,
}: GuestsRoomChildrenProps): ReactElement => {
  const dispatch = useDispatch();
  const {
    children,
    maxNumberOfChildren,
    maxOccupancy,
    totalGuests,
  }: RoomType = useSelector((state: StoreStateType) => state.rooms[position]);
  const isRemoveButtonDisabled = children.length === 0;
  const isAddButtonDisabled =
    children.length === maxNumberOfChildren || totalGuests === maxOccupancy;

  return (
    <>
      <Guest>
        <GuestText>Children</GuestText>
        <GuestButtonGroup>
          <Button
            onClick={() => dispatch(removeChild(position))}
            theme="secondary"
            dataTestId={`remove-child-${position}-button`}
            disabled={isRemoveButtonDisabled}
          >
            <IconMinus />
          </Button>
          <GuestText data-testid="children-counter">
            {children.length}
          </GuestText>
          <Button
            onClick={() => dispatch(addChild(position))}
            theme="secondary"
            dataTestId={`add-child-${position}-button`}
            disabled={isAddButtonDisabled}
          >
            <IconPlus />
          </Button>
        </GuestButtonGroup>
      </Guest>
      <KidsContainer>
        {children.map((child, index) => (
          <Guest key={`child-${index}-${position}`}>
            <GuestText>Child {index + 1} age</GuestText>
            <GuestButtonGroup>
              <Select
                dataTestId={`child-${index}-${position}-select`}
                value={String(child.age)}
                onChange={(event) =>
                  dispatch(
                    changeChildAge({
                      childrenPosition: index,
                      roomPosition: position,
                      age: event.target.value,
                    })
                  )
                }
              >
                <Option value="">Age</Option>
                {[
                  ...Array.from(Array(17).keys()).map((age) => (
                    <Option
                      key={`child-age-${age}-${position}-option`}
                      value={String(age)}
                    >
                      {age}
                    </Option>
                  )),
                ]}
              </Select>
              <IconRemoveChild
                data-testid={`remove-child-${index}-${position}-icon`}
                onClick={() =>
                  dispatch(
                    removeSpecificChild({
                      childrenPosition: index,
                      roomPosition: position,
                    })
                  )
                }
              />
            </GuestButtonGroup>
          </Guest>
        ))}
      </KidsContainer>
    </>
  );
};

export default GuestsRoomChildren;
