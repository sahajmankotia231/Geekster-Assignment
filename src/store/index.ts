import { createSlice, configureStore } from "@reduxjs/toolkit";

export type ChildType = {
  age: number | null;
};

export type RoomType = {
  maxNumberOfAdults: number;
  maxNumberOfChildren: number;
  maxOccupancy: number;
  adultsCounter: number;
  totalGuests: number;
  children: ChildType[];
};

export type StoreStateType = {
  maxNumberOfRooms: number;
  rooms: RoomType[];
};

const initialRoomsState = {
  maxNumberOfAdults: 5,
  maxNumberOfChildren: 3,
  maxOccupancy: 5,
  adultsCounter: 1,
  totalGuests: 1,
  children: [],
};

const initialState = {
  maxNumberOfRooms: 8,
  rooms: [initialRoomsState],
};

const counterSlice = createSlice({
  name: "rooms",
  initialState,
  reducers: {
    initializeStore: (state, action): void => {
      if (Boolean(action.payload)) {
        const splitedRooms = action.payload?.split("|") || [];
        const rooms = splitedRooms?.map(
          (room): RoomType => {
            const guests = room.split(":");
            const adultsCounter = Number(guests[0]);
            const children =
              guests[1]?.split(",")?.reduce((arr, age): ChildType[] => {
                arr.push({ age: Number(age) });
                return arr;
              }, []) || [];

            return {
              ...initialRoomsState,
              adultsCounter,
              children,
            };
          }
        );

        state.rooms = rooms;
      }
    },
    resetStore: (state): void => {
      state.rooms = [initialRoomsState];
    },
    addRoom: (state): void => {
      const { rooms, maxNumberOfRooms } = state;

      if (rooms.length < maxNumberOfRooms) {
        rooms.push(initialRoomsState);
      }
    },
    removeRoom: (state, action): void => {
      if (state.rooms.length > 1) {
        state.rooms.splice(action.payload, 1);
      }
    },
    addAdult: (state, action): void => {
      const room = state.rooms[action.payload];

      if (
        room.adultsCounter < room.maxNumberOfAdults &&
        room.totalGuests < room.maxOccupancy
      ) {
        room.adultsCounter += 1;
        room.totalGuests += 1;
      }
    },
    removeAdult: (state, action): void => {
      const room = state.rooms[action.payload];

      if (room.adultsCounter > 1) {
        room.adultsCounter -= 1;
        room.totalGuests -= 1;
      }
    },
    addChild: (state, action): void => {
      const room = state.rooms[action.payload];

      if (
        room.children.length < room.maxNumberOfChildren &&
        room.totalGuests < room.maxOccupancy
      ) {
        room.children.push({
          age: null,
        });
        room.totalGuests += 1;
      }
    },
    removeChild: (state, action): void => {
      const room = state.rooms[action.payload];
      const { children } = room;

      if (children.length) {
        children.splice(children.length - 1, 1);
        room.totalGuests -= 1;
      }
    },
    removeSpecificChild: (state, action): void => {
      const { childrenPosition, roomPosition } = action.payload;
      const room = state.rooms[roomPosition];
      const { children } = room;

      if (children.length) {
        children?.splice(childrenPosition, 1);
        room.totalGuests -= 1;
      }
    },
    changeChildAge: (state, action): void => {
      const { childrenPosition, roomPosition, age } = action.payload;
      const { children } = state.rooms[roomPosition];

      if (Boolean(age)) {
        children[childrenPosition].age = Number(age);
      }
    },
  },
});

export const {
  initializeStore,
  resetStore,
  addRoom,
  removeRoom,
  addAdult,
  removeAdult,
  addChild,
  removeChild,
  removeSpecificChild,
  changeChildAge,
} = counterSlice.actions;

const store = configureStore({
  reducer: counterSlice.reducer,
});

export default store;
