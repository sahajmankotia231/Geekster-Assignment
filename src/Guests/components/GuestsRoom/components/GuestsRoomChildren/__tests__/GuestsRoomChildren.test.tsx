import React from "react";
import { Provider } from "react-redux";
import { render, fireEvent, screen } from "@testing-library/react";

import store, { resetStore } from "store";

import GuestsRoomChildren from "..";

describe("<GuestsRoomChildren />", () => {
  beforeEach(() => store.dispatch(resetStore()));

  it("should render properly", () => {
    const { container } = render(
      <Provider store={store}>
        <GuestsRoomChildren position={0} />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should be able to add and remove a child", () => {
    render(
      <Provider store={store}>
        <GuestsRoomChildren position={0} />
      </Provider>
    );

    const childrenCounter = screen.getByTestId("children-counter");

    expect(childrenCounter.textContent).toBe("0");

    fireEvent.click(screen.getByTestId("add-child-0-button"));
    expect(childrenCounter.textContent).toBe("1");

    fireEvent.click(screen.getByTestId("remove-child-0-button"));
    expect(childrenCounter.textContent).toBe("0");
  });

  it("should be able to change a child age", () => {
    const childAge = 3;

    render(
      <Provider store={store}>
        <GuestsRoomChildren position={0} />
      </Provider>
    );

    fireEvent.click(screen.getByTestId("add-child-0-button"));
    fireEvent.change(screen.getByTestId("child-0-0-select"), {
      target: { value: childAge },
    });

    const child = store.getState().rooms[0].children[0];

    expect(child.age).toBe(childAge);
  });

  it("should disable remove button if the room has no children", () => {
    render(
      <Provider store={store}>
        <GuestsRoomChildren position={0} />
      </Provider>
    );

    expect(screen.getByTestId("remove-child-0-button")).toHaveAttribute(
      "disabled"
    );
  });

  it("should disable add button in room max of children", () => {
    const { maxNumberOfChildren } = store.getState().rooms[0];

    render(
      <Provider store={store}>
        <GuestsRoomChildren position={0} />
      </Provider>
    );

    const addButton = screen.getByTestId("add-child-0-button");

    expect(addButton).not.toHaveAttribute("disabled");

    for (let i = 1; i <= maxNumberOfChildren + 1; i++) {
      fireEvent.click(addButton);
    }

    expect(addButton).toHaveAttribute("disabled");
  });
});
