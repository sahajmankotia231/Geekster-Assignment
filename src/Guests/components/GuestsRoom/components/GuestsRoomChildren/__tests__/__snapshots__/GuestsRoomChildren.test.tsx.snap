// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GuestsRoomChildren /> should render properly 1`] = `
<div
  class="sc-gKckTs dPQnl"
>
  <span
    class="sc-iCfLBT JjZsB"
  >
    Children
  </span>
  <div
    class="sc-furvIG cWguzl"
  >
    <button
      class="sc-bdvvaa fzcnyp"
      data-testid="remove-child-0-button"
      disabled=""
    >
      <svg>
        minus.svg
      </svg>
    </button>
    <span
      class="sc-iCfLBT JjZsB"
      data-testid="children-counter"
    >
      0
    </span>
    <button
      class="sc-bdvvaa fzcnyp"
      data-testid="add-child-0-button"
    >
      <svg>
        plus.svg
      </svg>
    </button>
  </div>
</div>
`;
